core/templates/dev/head/components/SearchBarDirective.js

```diff
@@ -151,11 +167,68 @@ oppia.directive('searchBar', [function() {
             }
 
             _updateSelectionDetails('languageCodes');
+
+            if (Boolean(urlService.getUrlParams().q)) {
+              updateSearchFieldsBasedOnUrlQuery();
+            }
             _onSearchQueryChangeExec();
 
             _searchBarFullyLoaded = true;
           }
         );
+
+        var updateLanguageCodeFields = function(url) {
+          // Grabs Grab language code(s) from URL and returns url without
+          // language fields. This function assumes that language codes are
+          // at the end of the url.
+
+          var _languageCodePattern = / language_code=\([a-zOR\%\" ]+\)$/;
+          var _languageCodes = url.match(_languageCodePattern)[0];
+          var _languageCodeSelections = [];
+          if (_languageCodes != undefined) {
+            url = url.replace(_languageCodePattern, '');
+
+            // Grab language code(s) from extracted language section,
+            // these are 2 lowercase letters.
+            _languageCodeSelections = _languageCodes.match(/"([a-z]{2})"/g);
+          }
+
+          for (i = 0; i < _languageCodeSelections.length; i++) {
+            var language = _languageCodeSelections[i].match(/[a-z]{2}/g);
+            $scope.selectionDetails.languageCodes.selections[language] = true;
+          }
+          _updateSelectionDetails('languageCodes');
+          return url;
+        };
+
+        var updateCategoryFields = function(url) {
+          // Grab category code(s) from URL and returns the url without
+          // the category field.
+          var _categoryPattern = /category=\([A-Za-z0-9\" ]+\)\s*/;
+          var _categoriesCodes = url.match(_categoryPattern);
+
+          var _categories = [];
+          if (_categoriesCodes != undefined) {
+            url = url.replace(_categoryPattern, '');
+            _categories = _categoriesCodes[0].match(/"[A-Za-z]+"/g);
+          }
+
+          for (i = 0; i < _categories.length; i++) {
+            var category = _categories[i].match(/[A-Za-z]+/);
+            $scope.selectionDetails.categories.selections[category] = true;
+          }
+          _updateSelectionDetails('categories');
+          return url;
+        };
+
+        var updateSearchFieldsBasedOnUrlQuery = function() {
+          var query = decodeURIComponent(urlService.getUrlParams().q);
+          query = updateLanguageCodeFields(query);
```

Comment:
I still don't understand why we're mutating 'query', or passing the whole string to the individual functions. Instead, the query should be broken down in this function, and the individual pieces passed to the more specialized functions. The current design is not great because of the side-effects.